### PR TITLE
Form submit actions with HTTP POST

### PR DESCRIPTION
**Pull request by elpaso**
## New form submit POST actions

This PR adds two new HTTP POST actions, similar to the existing `Open` action which uses an HTTP GET request.

The two actions differ in the way the forma data are encoded:

1. application/x-www-form-urlencoded
2. multipart/form-data

form_submit_action

In both cases, data are passed to the action in an URL encoded URL.

A new function `url_encode(<map>)` is provided to convert a dictionary (a map) to the encoded format.

immagine

The response from the submitted URL is opened by delegating to the operating system `QDesktopServices::openUrl()` after the payload has been stored in a temporary file.

Funded by: Kanton Solothurn

### CODE REVIEW

File: src/core/qgsaction.cpp [hunk 1]
```diff
@@ -74,6 +87,183 @@ void QgsAction::run( const QgsExpressionContext &expressionContext ) const
...
+          const QString tempFilePath{ tempDir.path() + QDir::separator() + filename };
+          QFile tempFile{ tempFilePath };
+          tempFile.open( QIODevice::WriteOnly );
+          tempFile.write( replyData );
+          tempFile.close();
+          QDesktopServices::openUrl( QUrl::fromLocalFile( tempFilePath ) );
+        }
+        else
+        {
+          QgsMessageLog::logMessage( QObject::tr( "Redirect is not supported!" ), QStringLiteral( "Form Submit Action" ), Qgis::MessageLevel::Critical );
```
Comment (nyalldawson) [inline src/core/qgsaction.cpp:245]: Could we just use `req.setAttribute(QNetworkRequest::FollowRedirectsAttribute, true);` to delegate handling of this to Qt?

File: src/core/qgsaction.cpp [hunk 2]
```diff
@@ -74,6 +87,183 @@ void QgsAction::run( const QgsExpressionContext &expressionContext ) const
...
+      if ( reply->error() == QNetworkReply::NoError )
+      {
+
+        if ( reply->attribute( QNetworkRequest::RedirectionTargetAttribute ).isNull() )
+        {
+
+          const QByteArray replyData = reply->readAll();
+
+          QString filename { "download.bin" };
+          if ( const auto header = reply->header( QNetworkRequest::KnownHeaders::ContentDispositionHeader ).toString().toStdString(); ! header.empty() )
```
Comment (nyalldawson) [inline src/core/qgsaction.cpp:155]: there's so much auto use here that I can't review this code without pulling it into an IDE to actually see what's happening :sob:

File: src/core/qgsaction.cpp [hunk 3]
```diff
@@ -74,6 +87,183 @@ void QgsAction::run( const QgsExpressionContext &expressionContext ) const
     else
       QDesktopServices::openUrl( QUrl( expandedAction, QUrl::TolerantMode ) );
   }
+  else if ( mType == QgsAction::SubmitUrlEncoded || mType == QgsAction::SubmitUrlMultipart )
+  {
+
+    QUrl url{ expandedAction };
```
Comment (nyalldawson) [inline src/core/qgsaction.cpp:93]: would be nice to pull all this out to a separate function -- the ::run method is very long as a result!

File: resources/function_help/json/url_encode [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "url_encode",
+  "type": "function",
+  "groups": ["Maps"],
+  "description": "Returns an URL encoded string from a map.<br>The underlying function uses <a href='https://doc.qt.io/qt-5/qurlquery.html#toString'>QUrlQuery::toString</a> with <a href='https://doc.qt.io/qt-5/qurl.html#ComponentFormattingOption-enum'>QUrl::FullyEncoded</a> flags.<br>Note that the plus sign '+' is not converted.",
```
Comment (nyalldawson) [inline resources/function_help/json/url_encode:5]: Is the qt stuff relevant? This link is likely to break at some stage, so I'd rather see the content which is relevant for users included directly here. (Also it's not nice to direct end users to c++ docs, that's a big turn off for non-developers)

File: src/core/qgsaction.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/qgsaction.cpp:245]: We could, but redirect on POST is not a common HTTP workflow (it has potential security issues and browsers should prompt the user for confirmation), I'm not sure it is a wise move to support it.

File: src/core/qgsaction.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/qgsaction.cpp]: auto removed

File: resources/function_help/json/url_encode (hunk 4 above)
Comment (elpaso) [reply resources/function_help/json/url_encode]: done

File: src/core/qgsaction.cpp (hunk 3 above)
Comment (elpaso) [reply src/core/qgsaction.cpp]: done

File: src/core/qgsactionmanager.cpp [hunk 5]
```diff
@@ -197,6 +195,10 @@ void QgsActionManager::runAction( const QgsAction &action )
     else
       QDesktopServices::openUrl( QUrl( action.command(), QUrl::TolerantMode ) );
   }
+  else if ( action.type() == QgsAction::SubmitUrlEncoded || action.type() == QgsAction::SubmitUrlMultipart )
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/core/qgsactionmanager.cpp]: done

File: src/core/qgsaction.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsaction.cpp]: Fair enough!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**